File: hs/day6.hs
import qualified Data.ByteString as BS(readFile)
import Data.List
import qualified Data.Set as S
import Data.Text(Text)
import qualified Data.Text as T
import Data.Text.Encoding(decodeUtf8)

-- Bend space and time
-- https://stackoverflow.com/a/27733778/713961
windows :: Int -> [a] -> [[a]]
windows w = foldr (zipWith (:)) (repeat []) . take w . tails

uniqueWindowIndex :: Int -> Text -> Int
uniqueWindowIndex w input = windowIndex
    where
        windowIndex = case find unique indexed of
            Just (i, _) -> i + w
            Nothing -> error "no unique window"
        unique (_, wnd) = length (S.fromList wnd) == length wnd
        indexed = zip [0..] . windows w $ T.unpack input

part1 :: Text -> Int
part1 = uniqueWindowIndex 4

part2 :: Text -> Int
part2 = uniqueWindowIndex 14


main :: IO ()
main = do
    input <- decodeUtf8 <$> BS.readFile "../input/2022/day6.txt"
    print $ part1 input
    print $ part2 input

File: src/day6.rs
use aoc_runner_derive::aoc;

use rustc_hash::FxHashSet;

fn unique_window_index(input: &str, w: usize) -> usize {
    let (idx, _) = input.as_bytes().windows(w).enumerate().find(|(_, w)| {
        let uniq: FxHashSet<_> = w.iter().copied().collect();
        uniq.len() == w.len()
    }).expect("no unique window");
    idx + w
}

#[aoc(day6, part1)]
fn part1(input: &str) -> usize {
    unique_window_index(input, 4)
}

#[aoc(day6, part2)]
fn part2(input: &str) -> usize {
    unique_window_index(input, 14)
}
